services/file-retriever/src/services/dsnFetcher.ts

```diff
@@ -0,0 +1,99 @@
+import { FileResponse } from '../models/file.js'
+import {
+  blake3HashFromCid,
+  stringToCid,
+  decodeNode,
+  MetadataType,
+  cidToString,
+} from '@autonomys/auto-dag-data'
+import { PBNode } from '@ipld/dag-pb'
+import { env } from '../utils/env.js'
+import { HttpError } from '../http/middlewares/error.js'
+import { safeIPLDDecode } from '../utils/dagData.js'
+import mime from 'mime-types'
+
+const SUBSPACE_GATEWAY_URL = env('SUBSPACE_GATEWAY_URL')
+const MAX_SIMULTANEOUS_FETCHES = 10
+
+const fetchNode = async (cid: string): Promise<PBNode> => {
+  const objectMappingHash = blake3HashFromCid(stringToCid(cid))
+  const response = await fetch(
+    `${SUBSPACE_GATEWAY_URL}/objects/${objectMappingHash}`,
+  )
+  if (!response.ok) {
+    throw new HttpError(500, 'Internal server error: Failed to fetch node')
+  }
+
+  const blob = await response.arrayBuffer()
+
+  return decodeNode(blob)
+}
+
+const createReadStream = (node: PBNode): ReadableStream => {
+  const metadata = safeIPLDDecode(node)
+  if (metadata?.data) {
+    return new ReadableStream({
+      start: async (controller) => {
+        controller.enqueue(metadata.data)
+        controller.close()
+      },
+    })
+  }
+
+  return new ReadableStream({
```

Comment:
This method looks like a key piece in data reconstruction. Do you mind describing it?

Reply:
Sure